packages/contracts/contracts/zkopru/controllers/Migratable.sol

```diff
@@ -1,84 +1,127 @@
 // SPDX-License-Identifier: GPL-3.0-or-later
 pragma solidity =0.7.4;
+pragma experimental ABIEncoderV2;
 
 import { Storage } from "../storage/Storage.sol";
 import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 import { IERC721 } from "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+import { SafeMath } from "@openzeppelin/contracts/math/SafeMath.sol";
 import { MassDeposit, MassMigration, Types } from "../libraries/Types.sol";
 import { Deserializer } from "../libraries/Deserializer.sol";
+import { MerkleTreeLib, Hasher } from "../libraries/MerkleTree.sol";
+import { Hash } from "../libraries/Hash.sol";
 
 contract Migratable is Storage {
     using Types for *;
+    using MerkleTreeLib for Hasher;
+    using SafeMath for uint256;
 
     event NewMassMigration(
-        bytes32 checksum,
-        address network,
-        bytes32 merged,
-        uint256 fee
+        address sourceNetwork,
+        bytes32 migrationRoot,
+        bytes32 migrationHash
     );
 
-    /**
-     * @dev You can do the mass migration using this function. To execute
-     *      this function, the destination contract should inherits the
-     *      "Migratable" contract and have registered this current contract's
-     *      address as an allowed migrant.
-     * @param proposalChecksum Checksum of the block where the mass migration is included.
-     * @param // data Serialized mass migration data
-     */
-    function migrateTo(
-        bytes32 proposalChecksum,
-        bytes calldata //data
+    modifier onlyDestination(address dest) {
+        require(
+            msg.sender == dest,
+            "Only the destination contract can call this function."
+        );
+        _;
+    }
+
+    function migrateFrom(
```

Comment:
So to be clear, this function would be called once for Ether, and then once for each ERC20 token that exists in the contract? What about ERC721 support?

Reply:
That's correct. To migrate NFTs, users shouldd withdraw their NFT and deposit again for each NFTs.